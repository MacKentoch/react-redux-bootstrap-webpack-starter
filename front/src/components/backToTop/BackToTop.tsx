/* eslint-disable no-undefined */
import React, { useState, useEffect } from 'react';
import BackToTopButton from './backToTopButton/BackToTopButton';
import { Motion, spring, presets } from 'react-motion';

// #region types
type Props = {
  minScrollY: number;
  scrollTo?: string;
  onScrollDone?: () => any;
};
// #endregion

function BackToTop({ minScrollY = 120, onScrollDone }: Props) {
  const [showBackButton, setShowBackButton] = useState(false);
  const [windowScrollY, setWindowScrollY] = useState(0);
  const [tickingScollObserve, setTickingScollObserve] = useState(false);

  // #region on windows scroll callback
  const handleWindowScroll = () => {
    if (!window) {
      return;
    }

    /* eslint-disable no-undefined */
    const currentWindowScrollY =
      window.pageYOffset !== undefined
        ? window.pageYOffset
        : (
            document.documentElement ||
            document.body.parentNode ||
            document.body
          ).scrollTop;
    /* eslint-enable no-undefined */

    // scroll event fires to often, using window.requestAnimationFrame to limit computations
    if (!tickingScollObserve) {
      window.requestAnimationFrame(() => {
        if (windowScrollY !== currentWindowScrollY) {
          const shouldShowBackButton =
            currentWindowScrollY >= minScrollY ? true : false;

          setWindowScrollY(currentWindowScrollY);
          setShowBackButton(shouldShowBackButton);
        }
        setTickingScollObserve(false);
      });
    }

    setTickingScollObserve(true);
  };
  // #endregion

  // #region on button click (smooth scroll)
  const handlesOnBackButtonClick = (
    event: React.MouseEvent<HTMLButtonElement>,
  ) => {
    event && event.preventDefault();
    if (window && windowScrollY && windowScrollY > minScrollY) {
      // using here smoothscroll-polyfill
      window.scroll({ top: 0, left: 0, behavior: 'smooth' });
      typeof onScrollDone === 'function' && onScrollDone();
    }
  };
  // #endregion

  // #region mount and unmount subscrubstions
  useEffect(() => {
    if (typeof window !== 'undefined') {
      window.addEventListener('scroll', handleWindowScroll);
    }

    return function unsubscribeEvents() {
      if (typeof window !== 'undefined') {
        window.removeEventListener('scroll', handleWindowScroll);
      }
    };
  });
  // #endregion

  return (
    <Motion style={{ x: spring(showBackButton ? 0 : 120, presets.stiff) }}>
      {({ x }) => (
        <BackToTopButton
          position={'bottom-right'}
          onClick={handlesOnBackButtonClick}
          motionStyle={{
            WebkitTransform: `translate3d(${x}px, 0, 0)`,
            transform: `translate3d(${x}px, 0, 0)`,
          }}
        />
      )}
    </Motion>
  );
}

BackToTop.displayName = 'BackToTop';

export default BackToTop;
