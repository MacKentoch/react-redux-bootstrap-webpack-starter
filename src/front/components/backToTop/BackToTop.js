// @flow
/* eslint-disable no-undefined */

// #region imports
import React, { Component } from 'react';
import BackToTopButton from './backToTopButton/BackToTopButton';
import { Motion, spring, presets } from 'react-motion';
// #endregion

// #region flow types
type Props = {
  minScrollY: number,
  scrollTo?: string,
  onScrollDone?: () => any,
};

type State = {
  windowScrollY: number,
  showBackButton: boolean,
  tickingScollObserve: boolean,
};
// #endregion

class BackToTop extends Component<Props, State> {
  static defaultProps = {
    minScrollY: 120,
    onScrollDone: () => {},
  };

  state = {
    windowScrollY: 0,
    showBackButton: false,
    tickingScollObserve: false,
  };

  // #region lifecycle methods
  componentWillMount() {
    if (typeof window !== 'undefined') {
      window.addEventListener('scroll', this.handleWindowScroll);
    }
  }

  componentWillUnmount() {
    if (typeof window !== 'undefined') {
      window.removeEventListener('scroll', this.handleWindowScroll);
    }
  }

  render() {
    const { showBackButton } = this.state;
    return (
      <Motion style={{ x: spring(showBackButton ? 0 : 120, presets.stiff) }}>
        {({ x }) => (
          <BackToTopButton
            position={'bottom-right'}
            onClick={this.handlesOnBackButtonClick}
            motionStyle={{
              WebkitTransform: `translate3d(${x}px, 0, 0)`,
              transform: `translate3d(${x}px, 0, 0)`,
            }}
          />
        )}
      </Motion>
    );
  }
  // #endregion

  // #region on windows scroll callback
  handleWindowScroll = () => {
    if (window) {
      const { windowScrollY, tickingScollObserve } = this.state;
      const { minScrollY } = this.props;

      /* eslint-disable no-undefined */
      const currentWindowScrollY =
        window.pageYOffset !== undefined
          ? window.pageYOffset
          : (
              document.documentElement ||
              document.body.parentNode ||
              document.body
            ).scrollTop;
      /* eslint-enable no-undefined */

      // scroll event fires to often, using window.requestAnimationFrame to limit computations
      if (!tickingScollObserve) {
        window.requestAnimationFrame(() => {
          if (windowScrollY !== currentWindowScrollY) {
            const shouldShowBackButton =
              currentWindowScrollY >= minScrollY ? true : false;

            this.setState({
              windowScrollY: currentWindowScrollY,
              showBackButton: shouldShowBackButton,
            });
          }
          this.setState({ tickingScollObserve: false });
        });
      }

      this.setState({ tickingScollObserve: true });
    }
  };
  // #endregion

  // #region on button click (smooth scroll)
  handlesOnBackButtonClick = (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
    }
    const { minScrollY } = this.props;
    const { windowScrollY } = this.state;

    if (window && windowScrollY && windowScrollY > minScrollY) {
      // using here smoothscroll-polyfill
      window.scroll({ top: 0, left: 0, behavior: 'smooth' });
      // smoothScroll.scrollTo(scrollTo, this.scrollDone);
    }
  };

  scrollDone = () => {
    const { onScrollDone } = this.props;
    if (onScrollDone) {
      onScrollDone();
    }
  };
  // #endregion
}

export default BackToTop;
