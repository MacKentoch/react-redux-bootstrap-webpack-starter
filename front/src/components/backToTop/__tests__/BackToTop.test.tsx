import React from 'react';
import { shallow } from 'enzyme';
import BackToTop from '../BackToTop';

describe('BackToTop component', () => {
  it('renders as expected', () => {
    const props = {
      minScrollY: 10,
      onScrollDone: jest.fn(),
    };

    const wrapper = shallow(<BackToTop {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
