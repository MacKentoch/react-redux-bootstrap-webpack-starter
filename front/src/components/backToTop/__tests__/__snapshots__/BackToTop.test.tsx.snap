// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTop component renders as expected 1`] = `ShallowWrapper {}`;
